### PR TITLE
refactor package infrastructure/kubernetes

### PR DESCRIPTION
**Pull request by zirain**
refactor package to reuse code :
1. use `Applier` to create/update/delete Kubernetes resources
2. use `ResourceRender` to render resource for `Proxy`/`RateLimit` 
3. move common functions to `utils` package

cc @arkodg

### CODE REVIEW

File: internal/infrastructure/kubernetes/applier/apply.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			if err := i.Client.Update(ctx, cm); err != nil {
+				return fmt.Errorf("failed to update configmap %s/%s: %w", cm.Namespace, cm.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (i *Instance) DeleteConfigMap(ctx context.Context, cm *corev1.ConfigMap) error {
```
[the remark this answers is not in the document]
Comment (zirain) [reply internal/infrastructure/kubernetes/applier/apply.go:58]: IMO, this PR is too large to review, let's do more refactor later.
`Applier` absolute can be more generic.

**Comment by qicz** [comment]
> @qicz can you also help review this PR since you've done some refactoring in this area in the past, thanks in advance !

No problem. in fact, i have already reviewed a few days ago without.note, i will do this again.

File: internal/infrastructure/kubernetes/utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+// GetSelector returns a label selector used to select resources
+// based on the provided labels.
+func GetSelector(labels map[string]string) *metav1.LabelSelector {
+	return &metav1.LabelSelector{
+		MatchLabels: labels,
+	}
+}
+
+// expectedResourceHashedName returns hashed resource name.
+func ExpectedResourceHashedName(name string) string {
```
[the remark this answers is not in the document]
Comment (zirain) [reply internal/infrastructure/kubernetes/utils/utils.go:52]: it's used in both `proxy` and `ratelimit`

File: internal/infrastructure/kubernetes/proxy/resource_provider.go [hunk 3]
```diff
@@ -27,34 +28,148 @@ const (
...
-	// envoyHTTPPort is the container port number of Envoy's HTTP endpoint.
-	envoyHTTPPort = int32(8080)
-	// envoyHTTPSPort is the container port number of Envoy's HTTPS endpoint.
-	envoyHTTPSPort = int32(8443)
 )
 
-// expectedProxyDeployment returns the expected Deployment based on the provided infra.
-func (i *Infra) expectedProxyDeployment(infra *ir.Infra) (*appsv1.Deployment, error) {
+type ResourceRender struct {
+	infra *ir.Infra
```
[the remark this answers is not in the document]
Comment (zirain) [reply internal/infrastructure/kubernetes/proxy/resource_provider.go:34]: IMO, it's too wide to use `ProxyInfra` here for now, change it when necessary in the future.

**Review by qicz — APPROVED** [review]
LGTM, I will follow up for this refactor in another pr. issue #1330

File: internal/infrastructure/kubernetes/proxy/resource_provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zirain) [reply internal/infrastructure/kubernetes/proxy/resource_provider.go:34]: address follow up here: https://github.com/envoyproxy/gateway/issues/1330#issuecomment-1514186026

**Review by arkodg — APPROVED** [review]

**Pull request merged**